### DISCUSSION TITLE
send data to https webpage

### DISCUSSION THREAD

**Discussion by 2000ede** [arendst/Tasmota#17912] [Q&A] [answered]
Hello,
I found this command online "Rule1 on Energy#Power do var1 %value% endon on Time#Minute do WebSend [[192.168.178.200:8181](http://192.168.178.200:8181/)] /cuxd.exe?Status=dom.GetObject('solar').State(%var1%) endon"

I would also like to send something, but externally, i.e. to an https address. (google sheet)

everyone I've asked so far says it doesn't work with tasmota.
I do not understand that. if you set up an ESP8266, you can do it there too.
It would be a shame because it robs you of a lot of opportunities.
I would like to collect the data online without detours (mqtt).

is it really not possible? is there an adjustment that allows it?

Thanks for your tips.

**Comment by barbudor** [comment, marked as answer]
HTTPS is only supported on ESP32. ESP8266 is too light for that, only HTTP.
You should also look at WebQuery which syntax is more appropriate to address non-tasmota device
WebSend is more optimized to send http command to other tasmota device

If you must use an ESP8266, you will have to use some kind of external proxy rewrapper from HTTP to HTTPS
Could be Node-RED, probably NGINX, or even a few lines of Python

**Reply by barbudor** [reply]
To collect data, you can also use the available InfluxDB or Prometheus drivers
But that will be only in HTTP on ESP8266, not HTTPS

**Reply by 2000ede** [reply]
you may be right, but I'm pretty sure I sent values ​​from an ESP to Telegram a few years ago. Unfortunately I don't have it anymore.

I don't want to have any other devices locally apart from the Tasmota socket itself. It should be sent directly.

maybe this:
https://github.com/StorageB/Google-Sheets-Logging
(HTTPSRedirect)
- but i dont now in tasmota

**Reply by barbudor** [reply]
Telegram is a specific driver
general https is not supported on esp8266

**Reply by 2000ede** [reply]
what is this? HTTPSRedirect (https://github.com/StorageB/Google-Sheets-Logging)
(sorry, i am a Beginner)

**Reply by barbudor** [reply]
Ok, maybe I haven't been clear, so let's have the whole 3 keywords in the same sentence

HTTPS is not supported on TASMOTA for ESP8266

You can do it with Arduino with a sketch that does only what you need
Tasmota is doing a lot of stuff and we decided that it was not worth and we won't spend time on it
You are also free to fork and implement it by yourself

If you want HTTPS, move to ESP32
Especially if it's a DIY device, don't stick on ESP8266
ESP32C3 is as cheap as ESP8266 and full of better features

**Reply by 2000ede** [reply]
Thanks for your answer. I understand.

I don't want a bare ESP, that was just an explanation.

I want a WiFi socket with Tasmota. I was just wondering why it doesn't work there. A pity.

I don't tinker with ESP with 230 volts.

**Reply by barbudor** [reply]
Unfortunately there is very little plugs with ESP32 on the market today

Which plug standard are you looking for ? EU ? UK ? US ? AU ?
SwitchBot has one for the US market
I'm not aware any for EU market (and I don't follow other markets)

Alternatively there are devices like the recent Sonoff Elite POW16 and POW20
Or recent Shelly Pro devices
That needs cabling so a little more work than a plug

**Reply by 2000ede** [reply]
I am looking for EU plug or direct wiring. I know Sonoff, but I don't think it works with the original firmware either. I wanted to use Tasmota because it records the current data well.

**Reply by barbudor** [reply]
I was talking about devices that you can Tasmotize
Sonoff Elite POW16/20 can be flashed with Tasmota
And because ethey have an ESP32 you can use HTTPS
And you also have the powerfull Berry scripting language to do even more complex stuff directly on the device

**Reply by 2000ede** [reply]
thank you for having the patience to help me further.
So if it is an ESP32 model, then I can run the command without any further settings and problems? That would be great!

**Reply by barbudor** [reply]
Absolutely

**Reply by 2000ede** [reply]
another person wrote me that https sending with Tasmota will not work with an ESP32 either, because the WebSend command cannot do it

**Reply by sfromis** [reply]
Better be specific instead of unsourced and unsubstantiated claims.
```
12:49:39.773 CMD: WebQuery https://example.com/ GET
12:49:41.212 MQT: xxxxxxxx/stat/WEBQUERY = {"WebQuery":"Done"}
```

On ESP8266:
```
12:50:50.548 CMD: WebQuery https://example.com/ GET
12:50:50.819 MQT: xxxxxxxx/stat/WEBQUERY = {"WebQuery":"Connect failed"}
```

To be clear, WebQuery and WebSend is the same, except for WebSend using a less practical argument format, improved upon by creating WebQuery.

**Reply by barbudor** [reply]
If you asked to ChatGPT, don't be surprised if the answers

**Reply by sfromis** [reply]
Yes, ChatGPT would likely provide an answer with high confidence, also when it is just made up out of whole cloth, but such details does not seem to matter much :wink:

**Reply by sfromis** [reply]
Asked ChatGPT: "Does Tasmota running on ESP32 support https with the WebSend command?"
> No, Tasmota running on ESP32 does not support HTTPS with the WebSend command. The WebSend command is used to send data from Tasmota to a remote server over HTTP, and does not support encrypted HTTPS connections. If you need to securely send data from Tasmota to a remote server, you will need to use a different method that supports HTTPS, such as MQTT with a secured broker.

Except for being totally wrong about the support, the rest of the answer is not too bad :stuck_out_tongue_winking_eye:

**Reply by sfromis** [reply]
"Are you quite sure about that?"

> I apologize, it appears that I made a mistake in my previous answer. Yes, Tasmota running on ESP32 does support HTTPS with the WebSend command, provided that you have the necessary SSL certificates and configurations in place. The WebSend command in Tasmota allows you to send data to a remote server over an encrypted HTTPS connection, offering increased security for your data transmission.

**Reply by barbudor** [reply]
OMG, end of civilization is on its way
With such kind of Artificial Stupidiy to do their homework, the next generations will barely be able to feed themselves

**Reply by arendst** [reply]
what do we need kids for...

**Reply by sfromis** [reply]
ChatGPT is great for homework, with the task being to check out where ChatGPT got it wrong, and where it got it right :stuck_out_tongue_closed_eyes: 

The fun comes when ChatGPT is going to be used to answer that. Or maybe Bard (or whatever name Google is going to use).

**Comment by 2000ede** [comment]
I would like to test/try it again:
Remotely I now have a Tasmota system/socket based on an ESP32.

How do I enter the command correctly? When I enter this I get a message back "RSL: RESULT = {"WebSend":"Connect failed"}"

WebSend [https://script.google.com/macros/s/AKfycbwIVRUxkT_hZR3tq9_FSYdRcBLKQproqELwGby0S_KQn39O8H-3C5ygshPvImsiti-gng/exec?1_Spalte=100&2_Spalte=200&3_Spalte=300&4_Spalte=400&5_Spalte=500&6_Spalte=600]

(The URL commands via a browser works.)

Thanks for helping!

**Reply by sfromis** [reply]
You need to use Tasmota on ESP32 for that to work, not available on ESP8266.

If you have another device with Tasmota and ESP32, you could use that as a "bridge" to get things across, by triggering a WebSend command on that device.

**Reply by 2000ede** [reply]
yes, it's clear to me by now that it doesn't work on the ESP8266 directly.

That's why I tried it on the ESP32 (see entry). But it still doesn't work.

Sorry and thank you for your understanding for me as a beginner.

**Reply by barbudor** [reply]
The problem is somewhere else:
```
19:08:02.763 CMD: webquery https://script.google.com/macros/s/AKfycbwIVRUxkT_hZR3tq9_FSYdRcBLKQproqELwGby0S_KQn39O8H-3C5ygshPvImsiti-gng/exec?1_Spalte=100&2_Spalte=200&3_Spalte=300&4_Spalte=400&5_Spalte=500&6_Spalte=600 GET
19:08:03.605 MQT: stat/ble-bridge/WEBQUERY = {"WebQuery":"Done"}
```
And don't use WebSend. Better use WebQuery

**Reply by barbudor** [reply]
Most likely, your ESP32 doesn't have access to the internet
What is the result of `WebQuery https://www.google.com GET` ?

**Reply by 2000ede** [reply]
thanks, barbudor,

the ESP is on the internet (access the ESP via the internet, not at my house)

First:
20:01:03.841 CMD: webquery https://script.google.com/macros/s/AKfycbwIVRUxkT_hZR3tq9_FSYdRcBLKQproqELwGby0S_KQn39O8H-3C5ygshPvImsiti-gng/exec?1_Spalte=100&2_Spalte=200&3_Spalte=300&4_Spalte=400&5_Spalte=500&6_Spalte=600 GET

20:01:04.810 RSL: RESULT = {"WebQuery":"Done"}

unfortunately nothing arrived at google sheets (it can't be due to the https command itself, it arrives via the browser)

Secondly:
20:04:10.225 CMD: WebQuery https://www.google.com GET
20:04:11.075 RSL: RESULT = {"WebQuery":"Done"}

**Reply by barbudor** [reply]
Right before my previous message (18:09 GMT) I've called the given URL 3 times: 2 from my PC (curl) and 1 my ESP32 (webquery)
If it is working you should see my entries

**Reply by barbudor** [reply]
When you say it is working from your browser, is that from the same location as the ESP32 ?
Just done it 3 times again from my ESP32, do you see my entries ?

**Reply by barbudor** [reply]
Try from another browser than the one you are logged into your Google acount 
Is that working ?

**Reply by sfromis** [reply]
When trying it from the browser, you *may* be "piggybacking" on the browser already signed in to Google

**Reply by 2000ede** [reply]
I hadn't noticed that before. so crap.
yes, the command only comes from my PC - so you have to be logged in to google. I didn't recognize and consider that when I entered it in the browser. 
setback

**Reply by barbudor** [reply]
There are methods to get from google a signed URL that can be called from a non logged in client
That should be somewhere in the docs

**Reply by 2000ede** [reply]
I apologize again for overlooking that. I am a beginner.
The question is whether you can store a code like this in TASMOTA: 
https://esp32-server.de/google-sheets/

**Reply by sfromis** [reply]
I have no experience with this, but this thread suggests that you can add headers with POST for credentials when making requests.
https://github.com/arendst/Tasmota/discussions/15441

**Reply by 2000ede** [reply]
barbudor & sfromis, 
thank you for putting up with me for so long. Thanks to you I now have the solution!!!

I found the last error myself - it was in the webapp provision by google. I guess I was wrong here.

the solution is as the command structure from you, barbudor:
webquery https:/….. GET

THANKS!!!!!

**Comment by javier6d2** [comment]
Hello, I'm trying to make a POST with a ESP32  with HTTPS but it doesn't work. There is something wrong with the order? Same command using postman.co works correctly.

```
23:24:34.836 CMD: WebQuery https://api.smartthings.com/v1/devices/xxxxxxxx-802d-xxxx-8bed-d74a6283d5ed/commands POST [Authorization:Bearer xxxxxxx-fc6b-xxxx-a8eb-ff5698839dac\|Content-Type:application/json]{"message":"{"commands":[{"component":"main","capability":"switch","command":"on","arguments":[]}]}"}
23:24:35.554 RSL: RESULT = {"WebQuery":"Done"}
```

**Reply by sfromis** [reply]
Compared to the example linked from the docs about `WebQuery`, your format does look odd with `\|Content-Type` as that extra `\` before the `|` to delimit header fields seems to be something you've added. Doing that in a test, I saw the backslash being transmitted to the host, as part of the previous header field, which very plausibly would not be what it wants to get.

**Reply by javier6d2** [reply]
Thank you @sfromis, I didn't understood well this example in docs:

`WebQuery http://www.mysite.com/api/update POST [Authorization:Bearer xyz\|Content-Type:application/json]{"message":"body"}`: Sends POST data with an authorization header and Content-Type

 I thought extra` \ ` it was mandatory. Now with this new line code is working!

`WebQuery https://api.smartthings.com/v1/devices/xxxxxxxx-802d-xxxx-8bed-d74a6283d5ed/commands POST [Authorization:Bearer xxxxxxx-fc6b-xxxx-a8eb-ff5698839dac ] {"commands":[{"component":"main","capability":"switch","command":"on","arguments":[]}]}`

But with this other doesn't work:

`WebQuery https://api.smartthings.com/v1/devices/xxxxxxxx-802d-xxxx-8bed-d74a6283d5ed/commands POST [Authorization:Bearer xxxxxxx-fc6b-xxxx-a8eb-ff5698839dac|Content-Type:application/json] {"commands":[{"component":"main","capability":"switch","command":"on","arguments":[]}]}`

It seems I have some problem entering the second header.

**Reply by sfromis** [reply]
What I referred from the docs was this:
https://tasmota.github.io/docs/Commands/#webquery
Which links to this thread about the feature, without that extra `\`
https://github.com/arendst/Tasmota/pull/13209

Seems that there is a mess with the Commands page having two places giving information about `WebQuery`, and the erroneous `\` probably being due to trying to make docs layout work with the sometimes nebulous MakeDocs formatting markup. Fixed now.

With a different request, I tested that the format of `[Authorization:Bearer xxxxxxx-fc6b-xxxx-a8eb-ff5698839dac|Content-Type:application/json]` actually worked to have the server receive those two header fields in a normal header layout.

Without you telling what "doesn't work" vs "working" means, hard to speculate.

**Reply by javier6d2** [reply]
For those who do not have a Smartthings hub and cannot use the Tasmota Edge driver, here is a solution with ESP32:
[https://community.smartthings.com/t/smartthings-api-browser-now-available-to-all/249907](url)
Once the virtual switch device is created in Smartthings, we can turn it on or off using the WebQuery command of a Tasmota device.

`Rule1 ON POWER1#state==1 DO WebQuery https://api.smartthings.com/v1/devices/”ID device”/commands POST [Authorization:Bearer “your personal token”] {"commands":[{"component":"main","capability":"switch","command":"on"}]}
ENDON
`

`Rule2 ON POWER1#state==0 DO WebQuery https://api.smartthings.com/v1/devices/”ID device”/commands POST [Authorization:Bearer “your personal token”] {"commands":[{"component":"main","capability":"switch","command":"off"}]}
ENDON
`

Working means virtual Smartthing device goes on/off when you toggle Tasmota switch button. 
I was just curious why it doesn't work when I apply two headers `[Authorization:Bearer xxxxxxx-fc6b-xxxx-a8eb-ff5698839dac|Content-Type:application/json]` instead of one `[Authorization:Bearer xxxxxxx-fc6b-xxxx-a8eb-ff5698839dac]`

**Reply by sfromis** [reply]
As my testing indicates that Tasmota can correctly handle inserting two request headers, using the syntax you posted now, I'd say that what works or not is not a Tasmota question, probably more like up to what SmartThings expects. While it seems "surprising" if it fails when you add the plausible Content-Type header too, that looks like the SmartThings server not doing what you expect.
